core/src/main/java/fr/sncf/osrd/conflicts/Conflicts.kt

```diff
@@ -1,49 +1,170 @@
-            if (requirement.start < lastRequirement.end) {
-                res.add(Conflict(listOf(lastRequirement.trainId, requirement.trainId), requirement.start, lastRequirement.end, ConflictType.SPACING))
+        for (conflictGroup in detectRequirementConflicts(requirements) { _, _ -> true }) {
+            val trains = conflictGroup.map { it.trainId }
+            val beginTime = conflictGroup.minBy { it.beginTime }.beginTime
+            val endTime = conflictGroup.minBy { it.endTime }.endTime
+            res.add(Conflict(trains, beginTime, endTime, ConflictType.SPACING))
+        }
+    }
+    return res
+}
+
+
+interface RoutingTrainRequirement {
+    val trainId: Long
+    val routingRequirements: List<RoutingRequirement>
+}
+
+
+fun detectRoutingConflicts(trainsRequirements: List<RoutingTrainRequirement>): List<Conflict> {
+    val res = mutableListOf<Conflict>()
+
+    data class RoutingZoneConfig(val entryDet: String, val exitDet: String, val switches: Map<String, String>)
+    data class ZoneRequirement(
+        val trainId: Long,
+        val route: String,
+        override val beginTime: Double,
+        override val endTime: Double,
+        val config: RoutingZoneConfig,
+    ): ResourceRequirement
+
+    // reorganize requirements by zone
+    val zoneRequirements = mutableMapOf<String, MutableList<ZoneRequirement>>()
+    for (trainRequirements in trainsRequirements) {
+        val trainId = trainRequirements.trainId;
+        for (routeRequirements in trainRequirements.routingRequirements) {
+            val route = routeRequirements.route!!
+            val beginTime = routeRequirements.beginTime
+            for (zoneRequirement in routeRequirements.zones) {
+                val endTime = zoneRequirement.endTime
+                val config = RoutingZoneConfig(zoneRequirement.entryDetector, zoneRequirement.exitDetector, zoneRequirement.switches!!)
+                val requirement = ZoneRequirement(trainId, route, beginTime, endTime, config)
+                zoneRequirements.getOrPut(zoneRequirement.zone) { mutableListOf() }.add(requirement)
             }
         }
     }
 
+    // for each zone, check compatibility of overlapping requirements
+    for ((_, requirements) in zoneRequirements.entries) {
+        for (conflictGroup in detectRequirementConflicts(requirements) { a, b -> a.config != b.config }) {
+            val trains = conflictGroup.map { it.trainId }
+            val beginTime = conflictGroup.minBy { it.beginTime }.beginTime
+            val endTime = conflictGroup.minBy { it.endTime }.endTime
+            res.add(Conflict(trains, beginTime, endTime, ConflictType.ROUTING))
+        }
+    }
+    return res
+}
+
+interface ResourceRequirement {
+    val beginTime: Double
+    val endTime: Double
+}
+
+/**
+ * Return a list of requirement conflict groups.
+ * If requirements pairs (A, B) and (B, C) are conflicting, then (A, B, C) are part of the same conflict group.
```

Comment:
Didn't we change that  last PI? AFAIK the front expects the pairs (A, B) and (B, C). And, potentially, (A, C), if A and C are in directly in conflict